### PR TITLE
Increase fixedformat_in description size to 64K

### PR DESCRIPTION
**Pull request by csylvester-pivotal**
Increased the maximum length of the fixedwidth_in custom formatter
description from 8K to 64K. Allows for loading from fixed format files
into tables containing the maximum number of columns supported by
Greenplum (1600 as of 4.3.8.0).

### CODE REVIEW

**Comment by hlinnaka** [comment]
I'd suggest replacing the fixed buffer with a StringInfo. Makes the code more readable, as well as less error-prone.

**Comment by ivannovick** [comment]
@hlinnaka we had an old ticket about this issue and I think it mentioned something about being compatible with TOAST objects.  In any case do you see in any danger in this patch.  I know @csylvester-pivotal worked hard to get this working so if we can help him get it reviewed and merged that would be awesome!

**Comment by hlinnaka** [comment]
Well, I can't convince myself that 64k is indeed enough for all possible inputs. Maybe it is, maybe it isn't. It's not worth trying to find a maximum that's guaranteed to be large enough, because switching over to using a StringInfo makes the whole point moot, and makes the code shorter and more readable anyway.

I'm thinking something like the attached patch (completely untested, because I don't know how to reach this code
[pr-839-with-stringinfo.patch.txt](https://github.com/greenplum-db/gpdb/files/314379/pr-839-with-stringinfo.patch.txt)

)

**Comment by csylvester-pivotal** [comment]
Heikki,

Thanks for the help on this and the patch.

I have attached a test script I am using to test the fixed format
descriptor size. I've tested this with 64K and it passes.

I applied the patch you created and I'm getting a segmentation fault on one
of the segments. I'll have more time later today and tomorrow to try and
isolate what is happening but it has been years since I've done any serious
C dev :-)

But I like your solution of using StringInfo and will try and get it to
work. Regarding if 64K is enough, it will not cover all cases that I could
foresee. But my thoughts are 96K (or 128K at the most), would cover: max
columns supported in a table (1600) and max object name length (63 I think)
for column names.

Regards,
Craig

On Tue, Jun 14, 2016 at 12:05 PM, Heikki Linnakangas <
notifications@github.com> wrote:

> Well, I can't convince myself that 64k is indeed enough for all possible
> inputs. Maybe it is, maybe it isn't. It's not worth trying to find a
> maximum that's guaranteed to be large enough, because switching over to
> using a StringInfo makes the whole point moot, and makes the code shorter
> and more readable anyway.
>
> I'm thinking something like the attached patch (completely untested,
> because I don't know how to reach this code
> pr-839-with-stringinfo.patch.txt
> <https://github.com/greenplum-db/gpdb/files/314379/pr-839-with-stringinfo.patch.txt>
>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/839#issuecomment-225930713>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGO8dsa_6cG76vb9MKqb4WNNdEAuPgvxks5qLtE0gaJpZM4I1WTo>
> .
>



-- 


Craig Sylvester | Data Engineer | *Pivotal*

Phone:  703-362-0949 | csylvester@pivotal. <csylvester@gopivotal.com>io

**Comment by hlinnaka** [comment]
Ah, try replacing "while (key)" with "while (*key)". There's nothing like coding in the blind.. :-)

**Comment by csylvester-pivotal** [comment]
Hi Heikki,

Tested your changes and it worked great.

/Craig

On Tue, Jun 14, 2016 at 1:01 PM, Heikki Linnakangas <
notifications@github.com> wrote:

> Ah, try replacing "while (key)" with "while (*key)". There's nothing like
> coding in the blind.. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/839#issuecomment-225947071>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGO8dgUekaGCc_bT6SzSkbLyqCgM_qjPks5qLt5MgaJpZM4I1WTo>
> .
>



-- 


Craig Sylvester | Data Engineer | *Pivotal*

Phone:  703-362-0949 | csylvester@pivotal. <csylvester@gopivotal.com>io

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -17854,7 +17854,12 @@ static Datum transformFormatOpts(char formattype, List *formatOpts, int numcols,
 		/* custom format */
 		
 		int 		len = 0;
-		const int	maxlen = 8 * 1024 - 1;
+        /*
+         * Increase maxlen to 64K from 8K
```
Comment (ascherbaum-pivotal) [inline src/backend/commands/tablecmds.c]: This comment is correct, but wouldn't it be nicer to say "from 8K to 64K"? The way it is written now sounds like it's backwards.

**Comment by ascherbaum-pivotal** [comment]
If you increase the size because you want to fit all columns, it is helpful if you add MaxHeapAttributeNumber somewhere in the comment. This way, if someone searches for the maximum number columns in the source, she or he can spot this place.

**Comment by hlinnaka** [comment]
@ascherbaum-pivotal: I think we have consensus to switch to using StringInfo here instead of a fixed buffer, see the patch I attached to earlier comment. Could I entice you to pick up that patch and craft a suitable commit message and push that, instead? Or @csylvester-pivotal, did you plan to do that?

**Comment by ivannovick** [comment]
@ascherbaum-pivotal appreciate your help on this patch

**Comment by csylvester-pivotal** [comment]
I don't mind submitting a PR to use StringInfo instead of a fixed buffer
for the fixed format description. And Andreas I will note that this change
supports tables defined with the number of columns = MaxHeapAttributeNumber

/Craig

On Thu, Jun 16, 2016 at 2:00 PM, Ivan Novick <notifications@github.com>
wrote:

> @ascherbaum-pivotal <https://github.com/ascherbaum-pivotal> appreciate
> your help on this patch
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/839#issuecomment-226564069>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGO8dm2D6k1EZb0PwW2Jjc7r3EXeCXz_ks5qMY8sgaJpZM4I1WTo>
> .
>



-- 


Craig Sylvester | Data Engineer | *Pivotal*

Phone:  703-362-0949 | csylvester@pivotal. <csylvester@gopivotal.com>io

**Comment by danielgustafsson** [comment]
+1 on this squashed into a single commit

**Comment by hlinnaka** [comment]
Squashed and pushed. Thanks for the patch!

**Pull request closed without merging**